guix: Pointer Authentication and Branch Target Identification for aarch64 Linux

fanquake opened the pull request:
Arm Pointer Authentication (PAC) is a method of hardening code from Return Oriented Programming (ROP) attacks. It uses a tag in a pointer to sign and verify pointers. Branch Target Identification (BTI) is another code hardening method, where the branch/jump target is identified with a special landing pad instruction. Outside of some system support in glibc+kernel, packages gain the additional hardening by compiling with the `-mbranch-protection=flag` available in recent versions of GCC. In particular [-mbranch-protection=standard](https://gcc.gnu.org/onlinedocs/gcc-9.1.0/gcc/AArch64-Options.html) enables both BTI and PAC, with backwards compatible to `armv8.0` code sequences that activate on `v8.3` (PAC) & `v8.5` (BTI) enabled Arm machines. ([taken from Fedora](https://fedoraproject.org/wiki/Changes/Aarch64_PointerAuthentication)). 

Creation of a BTI enabled binary also requires that everything being linked in be BTI enabled. This means you currently cannot, for example, cross-compile using a Ubuntu based aarch64 toolchain, if you're wanting to use this feature. This can be shown using `-Wl,z,force-bti`, which will emit warnings for linked objects that are not BTI enabled (this is used in configure to detect when to disable using the flags). i.e:
```cpp
int main() { return 0; }
```

```bash
# aarch64-linux-gnu-g++ (Ubuntu 13.2.0-23ubuntu4) 13.2.0
aarch64-linux-gnu-g++ test.cpp -mbranch-protection=standard -Wl,-z,force-bti
/usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/bin/ld: /usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/lib/../lib/Scrt1.o: warning: BTI turned on by -z force-bti when all inputs do not have BTI in NOTE section.
```

Closes #19075.

hebasto commented:
Concept ACK.

From reading docs it's still unclear to me whether `-mbranch-protection=standard` implies `-mbranch-protection=bti`?


```
$ ./test/lint/lint-git-commit-check.sh 
The subject line of commit hash f799135959461079a220c5ddc97ea9a6b0056b2b is followed by a non-empty line. Subject lines should always be followed by a blank line.
```

luke-jr commented on configure.ac:
```diff
@@ -903,6 +903,20 @@ if test "$use_hardening" != "no"; then
       ;;
   esac
 
+  case $host in
+    *aarch64-unknown-linux*)
+      AX_CHECK_COMPILE_FLAG([-mbranch-protection=standard], [
+        SAVED_CXXFLAGS="$CXXFLAGS"
+        CXXFLAGS="$CXXFLAGS -mbranch-protection=standard"
+        AX_CHECK_LINK_FLAG([-Wl,-z,force-bti], [
+          HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -mbranch-protection=standard"
+          HARDENED_LDFLAGS="$HARDENED_LDFLAGS -Wl,-z,force-bti"
+        ], [], [$LDFLAG_WERROR])
+        AX_CHECK_LINK_FLAG([-Wl,-z,pac-plt], [HARDENED_LDFLAGS="$HARDENED_LDFLAGS -Wl,-z,pac-plt"], [], [$LDFLAG_WERROR])
+        CXXFLAGS="$SAVED_CXXFLAGS"
+      ])
+      ;;
+  esac
```
Won't this break a normal from-source build if the C/C++ standard libraries are BTI-ready, but other dependencies are not?

fanquake replied:
Made this Guix-only for now.

fanquake commented:
I've changed the approach here, and this is now based on #25437 and parts of #25484.

This adds `--enable-standard-branch-protection` to the configure flags when building GCC, which turns on the usage of branch-protection features by default. I've also added a commit that switches to using glibc 2.33 for the aarch64 build. glibc 2.32 was the first to ship with support for aarch64 branch protection features when built with a compatible compiler (see commit message for details). We couldn't actually use this for Guix builds, because it would break out back compat requires (the symbol checks fail), so this is still just for demonstration.

kristapsk commented:
Concept ACK

fanquake commented:
Rebased on master & #25861. Please review that PR first.

fanquake commented:
Rebased past #27029. Might split some more of this out.

fanquake commented:
Rebased past #27153.

fanquake commented:
Rebased past #27406.

Note that the branch protection option being added to libevent here, can now exist inside the `NO_HARDEN` clause.

fanquake commented:
Rebased onto #27897, which simplifies the actual changes here, and dropped no-longer needed commits.

fanquake committed:
guix: use glibc 2.33

glibc 2.32 was the first to ship with support for branch protection when
compiled with a compatible compiler, see below. However a number of
bugfixes/improvements shipped in glibc 2.33, so use that, rather than
trying to backport all relevant changes.

glibc 2.32 release notes: https://lwn.net/Articles/828210/

* AArch64 now supports standard branch protection security hardening
  in glibc when it is built with a GCC that is configured with
  --enable-standard-branch-protection (or if -mbranch-protection=standard
  flag is passed when building both GCC target libraries and glibc,
  in either case a custom GCC is needed).  This includes branch target
  identification (BTI) and pointer authentication for return addresses
  (PAC-RET).  They require armv8.5-a and armv8.3-a architecture
  extensions respectively for the protection to be effective,
  otherwise the used instructions are nops.  User code can use PAC-RET
  without libc support, but BTI requires a libc that is built with BTI
  support, otherwise runtime objects linked into user code will not be
  BTI compatible.

`__libc_single_threaded` added as it is now exported from at least
`bitcoin-wallet` and `test_bitcoin`.

fanquake commented:
Rebased on #29987.

fanquake commented:
Rebased on master and dropped a commit, also bumped the glibc 2.33 branch to the latest commit. Still based on #30433, but the main blocker here remains the glibc bump.

fanquake commented:
For now, the Guix built bins could be inspected with:
```bash
# bitcoin/guix-build-30af1c56da93/output/aarch64-linux-gnu/bitcoin-30af1c56da93/bin# readelf -n * | grep "AArch64"
      Properties: AArch64 feature: BTI, PAC
      Properties: AArch64 feature: BTI, PAC
      Properties: AArch64 feature: BTI, PAC
      Properties: AArch64 feature: BTI, PAC
      Properties: AArch64 feature: BTI, PAC
      Properties: AArch64 feature: BTI, PAC
      Properties: AArch64 feature: BTI, PAC
```

fanquake commented:
Rebased for #30433, and updated to add an export allowance for `__libc_single_threaded`.
